msal-go/design.md

```diff
@@ -0,0 +1,107 @@
+
+## General Structure
+
+Public Surface
+
+```bash
+apps/ - Contains all our code
+  confidential/ - The confidential application API
+  public/ - The public application API
+  cache/ - The cache interface that can be implemented to provide persistence cache storage of credentials
+```
+
+Internals
+
+```bash
+apps/
+  internal/
+    client/ - Shared package for common calls that Public and Confidential apps share
+    json/ - Our own json encoder/decoder for special needs
+    shared/ - Holds types that need to be in multiple packages and can't be moved into a single one due to import cycles
+    requests/ - The pacakge to communicate to services to get tokens
+```
+
+### Use of the Go special internal/ directory
+
+In Go, a directory called `internal/` contains packages that should only be used by other packages rooted at the same location. This is covered in the [official Go documentation](https://golang.org/doc/go1.4#internalpackages).
+
+For example, a package `.../a/b/c/internal/d/e/f` can be imported only by code in the directory tree rooted at `.../a/b/c`. It cannot be imported by code in `.../a/b/g` or in any other repository.
+
+We use this feature quite liberally to make clear what is using an internal package.  For example:
+
+```bash
+apps/internal/base - Only can be used by packages defined at apps/
+apps/internal/base/internal/storage - Only can be use by package client
+```
+
+## Public API
+
+The public API is encapsulated in `apps/`.  `apps/` has 3 packages of interest to users:
+
+- `public/` - Covers the Public Application Client. These are apps that run on the user's computer, such as command line interface apps. 
+- `confidential/` - Covers the Confidential Application Client. These are apps that run on servers - web apps, web APIs which get tokens for users and web APIs which get tokens on behalf of themselves (service to service communication)
+- `cache/` - This provides the interfaces that must be implemented to create persistent caches for any MSAL client
+
+## Internals
+
+In this section we will be talking about `internal/`.
+
+### JSON Handling
+
+JSON must be handled specially in our app. The basics are, if we receive fields that our structs do not contain, we cannot drop them, to ensure backwards and future compatibility.
+
+To handle that, we use our own custom `json` package that handles this.
+
+### Backend communication
+
+Communication to the backends is done via the `requests/` package. `oauth.Token` is the client for all communication.
+
+`oauth.Token` communicates via REST calls that are encapsulated in the `ops/` client.
+
+## Adding A Feature
+
+This is the general way to add a new feature to MSAL:
+
+- Add the REST calls to `ops.REST`
+- Add the higher level manipulations to `oauth.Token`
+- Add your logic to the `app/\<client\>` and access the services via your `oauth.Token`
+
+## Notable Differences To Other Clients
+
+### Confidential applications needs to handle multiple users without one big cache
```

Comment:
### Confidential applications need to handle multiple users without one big cache